datastore/txn.go

```diff
@@ -43,105 +43,116 @@ type Txn interface {
 
 	// OnDiscard registers a function to be called when the transaction is discarded.
 	OnDiscard(fn func())
+
+	// OnSuccessAsync registers a function to be called asynchronously when the transaction is committed.
+	OnSuccessAsync(fn func())
+
+	// OnErrorAsync registers a function to be called asynchronously when the transaction is rolled back.
+	OnErrorAsync(fn func())
+
+	// OnDiscardAsync registers a function to be called asynchronously when the transaction is discarded.
+	OnDiscardAsync(fn func())
 }
 
 type txn struct {
-	t ds.Txn
 	MultiStore
-
+	t  ds.Txn
 	id uint64
 
 	successFns []func()
 	errorFns   []func()
 	discardFns []func()
+
+	successAsyncFns []func()
+	errorAsyncFns   []func()
+	discardAsyncFns []func()
 }
 
 var _ Txn = (*txn)(nil)
 
-// NewTxnFrom returns a new Txn from the rootstore.
-func NewTxnFrom(ctx context.Context, rootstore ds.TxnDatastore, id uint64, readonly bool) (Txn, error) {
+func newTxnFrom(ctx context.Context, rootstore ds.TxnDatastore, readonly bool) (ds.Txn, error) {
 	// check if our datastore natively supports iterable transaction, transactions or batching
-	if iterableTxnStore, ok := rootstore.(iterable.IterableTxnDatastore); ok {
-		rootTxn, err := iterableTxnStore.NewIterableTransaction(ctx, readonly)
-		if err != nil {
-			return nil, err
-		}
-		multistore := MultiStoreFrom(ShimTxnStore{rootTxn})
-		return &txn{
-			rootTxn,
-			multistore,
-			id,
-			[]func(){},
-			[]func(){},
-			[]func(){},
-		}, nil
+	switch t := rootstore.(type) {
+	case iterable.IterableTxnDatastore:
+		return t.NewIterableTransaction(ctx, readonly)
+
+	default:
+		return rootstore.NewTransaction(ctx, readonly)
 	}
+}
 
-	rootTxn, err := rootstore.NewTransaction(ctx, readonly)
+// NewTxnFrom returns a new Txn from the rootstore.
+func NewTxnFrom(ctx context.Context, rootstore ds.TxnDatastore, id uint64, readonly bool) (Txn, error) {
+	rootTxn, err := newTxnFrom(ctx, rootstore, readonly)
 	if err != nil {
 		return nil, err
 	}
-
 	multistore := MultiStoreFrom(ShimTxnStore{rootTxn})
 	return &txn{
-		rootTxn,
-		multistore,
-		id,
-		[]func(){},
-		[]func(){},
-		[]func(){},
+		t:          rootTxn,
+		MultiStore: multistore,
+		id:         id,
 	}, nil
 }
 
-// ID returns the unique immutable identifier for this transaction.
 func (t *txn) ID() uint64 {
 	return t.id
 }
 
-// Commit finalizes a transaction, attempting to commit it to the Datastore.
 func (t *txn) Commit(ctx context.Context) error {
-	if err := t.t.Commit(ctx); err != nil {
-		runFns(t.errorFns)
-		return err
+	var fns []func()
+	var asyncFns []func()
+
+	err := t.t.Commit(ctx)
+	if err != nil {
+		fns = t.errorFns
+		asyncFns = t.errorAsyncFns
+	} else {
+		fns = t.successFns
+		asyncFns = t.successAsyncFns
 	}
-	runFns(t.successFns)
-	return nil
+
+	for _, fn := range fns {
+		fn()
+	}
+	for _, fn := range asyncFns {
```

Comment:
todo: The async functions need to be called before the non-async one otherwise they are less async than they could be..